### PR TITLE
[#261] Allow devices to have configuration templates

### LINKED ISSUE #261: Store a configuration file for each device
After a discussion with the PM the requirements have shifted a bit:

- Store a file containing the configuration of each device (not each decoder). The format of the file is not assumed to be JSON, it could be any kind of file, and we will not perform any validation on it.
- We will not store a configurationSchema since we will allow configurations to have any file format. We will leave it up to the device to reject a configuration if it does not like it.
- Allow devices to register without providing a starting configuration.
- Allow users to call an endpoint to upload a new configuration file which will replace the old one.
- Allow users to call an endpoint to retrieve the current configuration
- Track the time at which the configuration was last modified. Make sure this value cannot be "manually" changed, it should be set to the current time whenever the config is changed.
- Store the config as a binary file in the db to disallow database injection.

### PR DESCRIPTION
**Pull request by mohhef**
# General info
1. The task is done on a separate endpoint because, in spring boot multipartfile, it is not allowed to upload a file + have a request body
2. I save files with bytes, thus should not be able to do db attacks

3. to test this pr
image

### CODE REVIEW

File: src/main/java/org/beanpod/switchboard/controller/EncoderController.java
Comment (felixlapierre) [inline src/main/java/org/beanpod/switchboard/controller/EncoderController.java]: Rename method

File: src/main/java/org/beanpod/switchboard/controller/DeviceController.java
Comment (felixlapierre) [inline src/main/java/org/beanpod/switchboard/controller/DeviceController.java]: Change to "Configuration uploaded." since there is only one file now.

File: src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java
Comment (felixlapierre) [inline src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java]: Rename, since we aren't doing exclusively json anymore

File: src/test/java/org/beanpod/switchboard/fixture/DeviceFixture.java
Comment (felixlapierre) [inline src/test/java/org/beanpod/switchboard/fixture/DeviceFixture.java]: Rename, since the type doesn't have to be json anymore. Value is probably fine though.

File: src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java
Comment (felixlapierre) [inline src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java]: Should assert somewhere that the device being saved in `deviceService` actually has the correct `configurationInstance`

File: contracts/api-spec.yaml
Comment (felixlapierre) [inline contracts/api-spec.yaml]: We should just rename this to `configuration` since we don't need a distinction between instance and schema anymore.

**Review by felixlapierre — COMMENTED** [review]
Mostly suggesting renames here. Also, we should probably wait until restrict access is merged before merging this.

File: contracts/api-spec.yaml
Comment (mohhef) [reply contracts/api-spec.yaml]: fixed

File: src/test/java/org/beanpod/switchboard/fixture/DeviceFixture.java
Comment (mohhef) [reply src/test/java/org/beanpod/switchboard/fixture/DeviceFixture.java]: fixed

File: src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java
Comment (mohhef) [reply src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java]: fixed

File: src/main/java/org/beanpod/switchboard/controller/DeviceController.java
Comment (mohhef) [reply src/main/java/org/beanpod/switchboard/controller/DeviceController.java]: fixed

File: src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java
Comment (mohhef) [reply src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java]: how would you verify that?

Comment (felixlapierre) [reply src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java]:
```
DeviceDto expectedSavedDevice = DeviceFixture.getDeviceDto();
expectedSavedDevice.setConfigurationInstance(validJsonFile.getBytes());
verify(deviceService).save(expectedSavedDevice);
```

**Review by felixlapierre — DISMISSED** [review]

File: src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java
Comment (mohhef) [reply src/test/java/org/beanpod/switchboard/controller/DeviceControllerTest.java]: Fixed

**Review by sebhani — DISMISSED** [review]
Works fine!

**Review by felixlapierre — APPROVED** [review]

**Review by sebhani — APPROVED** [review]

**Pull request merged**
